### PR TITLE
Ml with brainjs

### LINKED ISSUE #2572: Getting started with JavaScript Machine Learning using Brain.js
### Proposed title of article
Getting started with JavaScript Machine Learning using Brain.js

### Introduction paragraph (2-3 paragraphs):
Brain.js is a JavaScript Library used to building Neural networks in JavaScript. Brain.js is easy and simple to use.
In this tutorial, the readers will understand how to create a neural network, add training data and  test a neural network form scratch using brain.js.
### Key takeaways: 
1. What is brain.js?
2. Getting started with brain.js
3. Building, training and testing a neural network with brain.js.

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by jamila-laureen**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #2572

### CODE REVIEW

**Comment by jerimkaura** [comment]
Hello @jamila-laureen ,

Welcome to Section's Eng-Ed program, thank you for submitting your first article for review. 
 
I will take you through the review beginning with preliminary revision where I will check your article for grammatical correctness.

File: content/articles/ml-with-brainjs/index.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
+### Introduction
+Everybody knows that machine learning is mathematically intensive. For most beginners trying to get into machine learning, mathematical knowledge is a challenge.However, Brain.js makes the task easier, as it completely hides the mathematical logic and complexity that comes with understanding neural networks. It is therefore easy for beginners to get started with machine learning.
+
+### Understanding Neural Networks
+You can think of a neural network like the human brain, with several interconnected nodes grouped into layers. The layers are categorized as input, hidden, and output layers. Data in neural networks move in a single direction from the input towards the output.
```
Comment (jerimkaura) [inline content/articles/ml-with-brainjs/index.md:5]: Neural network is noun so capitalize its first letter

File: content/articles/ml-with-brainjs/index.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+### Project Goal
+The main goal of this article is to provide beginners, especially those who are coming from a web development background, to get in machine learning. In the process, we will build a neural network, give it a set of data to train on, then ensure that our model can make a prediction based on the learning set of data provided. 
+
+The project will be based on the JavaScript machine learning library called [Brain.js](https://brain.js.org/#/).
+
+### Project Setup
+Run `npm init -y` to create an empty `package.json` file. Next, execute the command `npm install brain.js` in your terminal to install the Brain.js package. In the root folder of the application, we will have the `index.js` file, which contains the driver called for application. Will also have a file that contains the training data for our neural network. The training data will be in the form of JSON, so we will create a `data.json` file. The final project structure should look like this:
```
Comment (jerimkaura) [inline content/articles/ml-with-brainjs/index.md:28]: Make this statement clear enough

File: content/authors/jamila-laureen/index.md [hunk 3]
```diff
@@ -0,0 +1 @@
+My name is Jamila Laureen, I am an undergraduate student persuing Infomation Technology. I love programming and Designing. At my free time, I like bike riding.
```
Comment (jerimkaura) [inline content/authors/jamila-laureen/index.md:1]:
```suggestion
---
title: Jamila Laureen
type: authors
---
Jamila Laureen is an undergraduate student pursuing Infomation Technology. She loves programming and Designing. At her free time, she likes bike riding.
```

**Review by jerimkaura — CHANGES_REQUESTED** [review]
@jamila-laureen, Your article flows just fine.:rocket:.Kindly take a look at the suggestions and ping me once you have done the changes.

**Comment by jamila-laureen** [comment]
@jerimkaura, I have made the changes.

**Comment by jerimkaura** [comment]
@jamila-laureen, I will proceed to the round of tech review where I test your code snippets.

**Review by jerimkaura — APPROVED** [review]

**Review by jerimkaura — CHANGES_REQUESTED** [review]
@jamila-laureen, Hello, I hope you are well.

Your article is amazing, I tested the snippets and they work well. However, Would you consider adding a table of content to your article?

If you have any issues adding the `TOC` then ping me we work on it together.

**Review by jerimkaura — APPROVED** [review]

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/ml-with-brainjs/index.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+/**
+ * Create the  neural network
+ */
+const network = new brain.recurrent.LSTM();
+```
+
+### Training the Model
+To train the model, we will take the data from our data file, convert it into an array of values with input and output pairs. We will train our model by running several interactions through the data pair. 
+
+This procedure will take time depending on the number of iteration you perform on the dataset. The more the number of iterations, the higher the accuracy and time it takes.
```
Comment (ahmadmardeni1) [inline content/articles/ml-with-brainjs/index.md]: This is not correct/accurate. @jamila-laureen 
If the learning rate is small enough, then training a neural network with too many epochs will result in loss converging to one of its local minimum since at every epoch we are moving towards minimum. Once a minimum is reached, then further training will not change the loss since the parameters are updated as Θ←Θ−α∇ΘL and  ∇ΘL=0  at minimum. In other words, the parameters  Θ  does not change.

However, in most cases, we do not want the loss to reach its minimum value since that would result in a neural network which is more prone to over-fitting. The loss function is a measure of how well the neural network does on the training set. If the loss is minimised beyond a certain point, then the neural network is likely to do worse on the unseen examples, i.e., the neural network will have poor generalization.

In other words, as the number of epochs increases, at first both the training error and the test error start decreasing, but after a certain point, the test error starts to increase (the training error continues to decrease since we are minimizing the training error). You can observe this in the following plot of learning curves:



The main goal of training a neural network is to do well on unseen examples, not on training examples. That is, we want the neural network to have good generalization.

**Comment by ahmadmardeni1** [comment]
Welcome to Section EngEd and congrats on your first tutorial 🎉 @jamila-laureen Please go over the change I made and the comment above. 

Thank you for the review @jerimkaura
